venntdb.py

```diff
@@ -2,110 +2,110 @@
+    from db_initiative import reset_turn_order, add_turn, next_turn, get_turn_order, get_current_turn
+    from db_combat import push_undo, pop_undo, get_undo_history
+
+    def __init__(self, filename):
+        self.filename = filename
+        self.auth = Authenticator()
+
+        if os.path.exists(self.filename):
+            self.db = cPickle.load(open(self.filename, 'rb'))
+        else:
+            self.db = {}
+            self.db["accounts"] = {}
+            self.db["campaigns"] = {}
+            self.db["weapons"] = {}
+            if os.path.exists("weapons.json"):
+                with open("weapons.json") as f:
+                    self.db["weapons"] = json.load(f)
+            self.db["abilities"] = {}
+            self.db["ability_cache"] = [None] * \
+                (MAX_ABILITY_CACHE + 1)  # list of Ability dicts
+            self.db["ability_cache_index"] = 0
+            # TODO: Do not include static abilties json files in the database for efficiency
+            if os.path.exists("abilities.json"):
+                with open("abilities.json", encoding="utf8") as f:  # encoding for smart quotes
+                    self.db["abilities"] = json.load(f)
+                    # self.validate_abilities() # TODO continue
+
+    def dump(self):
+        print(json.dumps(self.db, indent=4, separators=(',', ': '), sort_keys=True))
+
+    def permissions(self, username, character_id):
+        # TODO Permission.EDIT unused
+        if self.is_valid("accounts", username, "characters", character_id):
+            return Permission.OWN
+        else:
+            for campaign_id in self.db["accounts"][username]["joined_campaigns"]:
+                campaign = self.db["campaigns"][campaign_id]
+                role = campaign["members"][username]
+                if role == "GM":
+                    if character_id in campaign["members"].keys():
+                        return Permission.PRIVATE_VIEW
+                elif role == "player":
+                    if character_id in campaign["members"].keys():
+                        return Permission.ADD  # maybe COMBAT?
+                else:
+                    if character_id in campaign["members"].keys():
+                        return Permission.PUBLIC_VIEW
+
+        return Permission.NONE
+
+    def assert_valid(self, *args, dict=None):
+        logger.log("assert_valid", ">".join(args))
+        if not self.is_valid(*args, dict=dict):
+            raise AssertionError("Database path does not exist")
+
+    def is_valid(self, *args, dict=None):  # check valid path through DB
+        if dict is None:
+            dict = self.db
+        if len(args) < 1:
+            return True
+        elif len(args) == 1:
+            return args[0] in dict
+        else:
+            return args[0] in dict and self.is_valid(*args[1:], dict=dict[args[0]])
+
+    def create_account(self, username, pass_hash):
```

Comment:
Tracking DB changes: removed the `joined_campaigns` section - instead we just use the `campaigns` section